baremetal/metal3machine_manager.go

```diff
@@ -522,60 +564,66 @@ func (m *MachineManager) Delete(ctx context.Context) error {
-			}
-		}
-		if m3mt != nil {
-			if m3mt.Spec.NodeReuse {
-				if host.Labels == nil {
-					host.Labels = make(map[string]string)
+		if m.Cluster != nil {
+			// If cluster has DeletionTimestamp set, skip checking if nodeReuse
+			// feature is enabled.
+			if m.Cluster.DeletionTimestamp.IsZero() {
+				// Fetch corresponding Metal3MachineTemplate, to see if nodeReuse
+				// feature is enabled. If set to true, check the machine role. In case
+				// machine role is ControlPlane, set nodeReuseLabelName to KubeadmControlPlane
+				// name, otherwise to MachineDeployment name.
+				m.Log.Info("Getting Metal3MachineTemplate")
+				m3mt := &capm3.Metal3MachineTemplate{}
+				if m.Metal3Machine == nil {
+					return errors.New("Metal3Machine associated with Metal3MachineTemplate is not found")
 				}
-				// Check if machine is ControlPlane
-				if m.isControlPlane() {
-					// Fetch KubeadmControlPlane name for controlplane machine
-					m.Log.Info("Fetch KubeadmControlPlane name")
-					kcpName, err := m.getKubeadmControlPlaneName(ctx)
-					if err != nil {
-						return err
+				if m.hasTemplateAnnotation() {
+					m3mtKey := client.ObjectKey{
+						Name:      m.Metal3Machine.ObjectMeta.GetAnnotations()[capi.TemplateClonedFromNameAnnotation],
+						Namespace: m.Metal3Machine.Namespace,
 					}
-					m.Log.Info("Fetched KubeadmControlPlane name:", "kubeadmcontrolplane", kcpName)
-					// Set the nodeReuseLabelName to KubeadmControlPlane name on the host
-					m.Log.Info("Setting nodeReuseLabelName in BaremetalHost to fetched KubeadmControlPlane name")
-					host.Labels[nodeReuseLabelName] = kcpName
-				} else {
-					// Fetch MachineDeployment name for worker machine
-					m.Log.Info("Fetch MachineDeployment name")
-					mdName, err := m.getMachineDeploymentName(ctx)
-					if err != nil {
-						return err
+					if err := m.client.Get(ctx, m3mtKey, m3mt); err != nil {
+						// we are here, because while normal deprovisioning, Metal3MachineTemplate will be deleted first
+						// and we can't get it even though Metal3Machine has reference to it. We consider it nil and move
+						// forward with normal deprovisioning.
+						m3mt = nil
+						m.Log.Info("Metal3MachineTemplate associated with Metal3Machine is deleted")
+					} else {
+						// in case of upgrading, Metal3MachineTemplate will not be deleted and we can fetch it,
+						// in order to check for node reuse feature in the next step.
+						m.Log.Info(fmt.Sprintf("Found Metal3machineTemplate %v", m3mtKey.Name))
+					}
+				}
+				if m3mt != nil {
+					if m3mt.Spec.NodeReuse {
+						if host.Labels == nil {
+							host.Labels = make(map[string]string)
+						}
+						// Check if machine is ControlPlane
+						if m.isControlPlane() {
+							// Fetch KubeadmControlPlane name for controlplane machine
+							m.Log.Info(fmt.Sprintf("Fetch KubeadmControlPlane name while deprovisioning host %v", host.Name))
```

Comment:
Just nit: This seems a bit redundant and we have quite many logging in this if already.

Reply:
I think this is helpful to know which BMH being deprovisioned and in which BMH we expect nodeReuse label while debugging